File: leetcode/easy/ConstructStringFromBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public String tree2str(TreeNode t) {
        if (t == null) {
            return "";
        }
        StringBuilder sb = new StringBuilder(String.valueOf(t.val));
        // left
        if (t.left != null) {
            sb.append('(').append(tree2str(t.left)).append(')');
        } else if (t.right != null) {
            // left cannot be ommitted if right exists
            sb.append("()");
        }
        // right
        if (t.right != null) {
            sb.append('(').append(tree2str(t.right)).append(')');
        }
        return sb.toString();
    }
}

File: leetcode/easy/ConstructStringFromBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def tree2str(self, t):
        """
        :type t: TreeNode
        :rtype: str
        """
        if not t:
            return ''
        s = str(t.val)

        if t.left:
            s += '(' + self.tree2str(t.left) + ')'
        elif t.right:
            # left cannot be ommitted if right exists
            s += '()'

        if t.right:
            s += '(' + self.tree2str(t.right) + ')'
        return s;

